### PR TITLE
Improve tehuti-vis plotting

### PR DESCRIPTION
**Pull request by dkillick**
Improve the tehuti-vis plotting functionality, primarily so that any number of commits can be plotted in a single plot.

The plotting module has been re-designed to follow the OO state pattern so that new plotting methods (states) can be easily added. State classes are included in the new file `vis_methods`.

### CODE REVIEW

File: tehuti-vis.py [hunk 1]
```diff
@@ -17,78 +17,95 @@
...
+        self.results = results
+        self._method = method
+        self.plot_data = None
+
+        self.methods = {'basic': VaryRepoCommit(self),
+                        'violin': Violin(self)}
+
+    @property
+    def method(self):
+        if self._method is None:
```
Comment (bjlittle) [inline tehuti-vis.py]: @dkillick How can this be `None`? I'm guessing only if the instance has been instantiated explicitly by the user with a call like `Vis(results, None)` ... seems confusing. Does this mean that `method` in the `__init__` is a kwarg? I'm thinking not ... 

How about removing this check in here and putting it in the `__init__` ?

File: tehuti-vis.py [hunk 2]
```diff
@@ -17,78 +17,95 @@
...
-            width = 0.6
-            pos = [x - (width / 2) for x in range(len(data))]
-            ax.bar(pos, data, width=width)
-        ax.set_xticks(range(len(data)))
-        ax.set_xticklabels(labels, rotation=30)
-        plt.tight_layout()
-        plt.show()
+            self._method = self.methods[value]
+        except KeyError:
+            msg = 'Vis has no method {}.'
```
Comment (bjlittle) [inline tehuti-vis.py]: @dkillick Use `{!r}` perhaps ...

File: tehuti-vis.py [hunk 3]
```diff
@@ -17,78 +17,95 @@
...
+            raise AttributeError(msg.format(value))
+
+    def select_data(self, commits=None, metrics=None):
+        """
+        Select data to plot. Processing is handed off to the `select_data`
+        method of the current state.
+
+        Args:
```
Comment (bjlittle) [inline tehuti-vis.py]: @dkillick These are `Kwargs` ...

File: tehuti-vis.py (hunk 1 above)
Comment (dkillick) [inline tehuti-vis.py]: Good spot - I think this is a hangover from an earlier approach to getting and setting the method. The check isn't actually needed any more!

File: tehuti-vis.py [hunk 4]
```diff
@@ -17,78 +17,95 @@
...
+        Construct a tehuti metrics visualiser.
+
+        The data selected and plots produced are determined by the
+        visualisation method (or state) that the visualiser is constructed
+        with.
+
+        Args:
+
+        * results:
+            The benchmarking results to visualise.
```
Comment (bjlittle) [inline tehuti-vis.py]: @dkillick Any hints as to what this `results` thing is?

File: tehuti-vis.py [hunk 5]
```diff
@@ -17,78 +17,95 @@
...
+
+        Arg:
+
+        * alternate_plot:
+            Toggle to select the alternate plotting mode of the `plot` method
+            of the current state.
+
+        """
+        if self.plot_data is None:
+            self.select_data()
```
Comment (bjlittle) [inline tehuti-vis.py]: @dkillick Just a thought, but if `self.plot_data is None` then I guess the intention is that `self.select_data()` will populate `self.plot_data` ... but will this always be the case? Given that the call `self.select_data` let's it kwargs default to None.

File: vis_methods.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+class VaryRepoCommit(Visualiser):
+    """
+    A visualiser state that plots benchmark results against repository commits
+    for all supplied metrics.
+
+    Metric results may either be plotted each on individual axes, or all on the
+    same axes.
+
+    """
+    def __init__(self, vis):
```
Comment (bjlittle) [inline vis_methods.py]: @dkillick Apologies, but doc-string ...

File: vis_methods.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+            One or more valid repository commits that have been benchmarked.
+        * metrics:
+            One or more metrics that have results in the metrics results file.
+
+        Returns:
+            The selected data formatted for plotting.
+
+        """
+        commits, metrics = super(VaryRepoCommit, self).select_data(commits,
+                                                                   metrics)
```
Comment (bjlittle) [inline vis_methods.py]: @dkillick Oh, be careful here with this pattern ... this `select_data` is specializing with the same signature as its parent, but they return different things i.e. the parent returns a tuple of `commits, metrics` whereas the sub-class returns just `data` ... you may want to rename the common parent method? It's a subtle difference ...

File: tehuti-vis.py (hunk 5 above)
Comment (dkillick) [inline tehuti-vis.py]: Yes, `self.select_data()` should always populate `self.plot_data`, but it doesn't _have_ to be called (though in practice it is likely to be as this is all designed to be run from the `__main__` section of this module).
Not setting the kwargs for `self.select_data()` actually means that all metrics and commits found in the results file are selected to be plotted, but this wasn't clear from it's docstring. I've updated the docstring to make this clearer.

File: vis_methods.py (hunk 6 above)
Comment (dkillick) [inline vis_methods.py]: What do you mean..?

Comment (bjlittle) [inline vis_methods.py]:
```python
def __init__(self, vis):
    """
    Blah blah ...

    Args:

    * vis:
        This is a vis do-dah-thingy ...

    """
```

Comment (dkillick) [inline vis_methods.py]: Oh, for the `__init__`! I was like, "there already is a docstring!"...

Comment (bjlittle) [inline vis_methods.py]: :wink:

**Comment by dkillick** [comment]
Review actions: tidied some bits and pieces as per @bjlittle's suggestions.

**Comment by bjlittle** [comment]
@dkillick Great stuff, thanks! :+1:

**Pull request merged**